Replay progress events in unit tests

gilest opened the pull request:
Skips the browser, http, network etc. and lets us test the smallest unit of our internals – an `UploadFile` having its state updated by a series of `ProgressEvent` events.

For each scenario I'm pretty sure we only need to test a single file with a stream of synchronous events.

Covers some recorded events from different browsers and endpoints. We can add more scenarios here if we come across them.

Specifically, the Chrome scenario proves the fix in #1016 – this was previously failing early with 100% progress, which is the same issue reported by users in #1013

I've painstakingly copied values from log screenshots to setup this existing suite. Hopefully contributors can PR full tests in future to make things a bit easier 😅 

Fixes #1013

<img width="960" alt="image" src="https://github.com/adopted-ember-addons/ember-file-upload/assets/36919/2ad3aaa1-a3b7-4b4f-aed9-d8342f43f526">

RobbieTheWagner commented:
Thanks for the quick turnaround on this @gilest! Excited to see this come together 😄

RobbieTheWagner commented:
Thanks for the quick turnaround on this @gilest! Excited to see this come together 😄

gilest commented on test-app/tests/unit/system/upload-progress-test.ts:
```diff
@@ -0,0 +1,202 @@
+/* eslint-disable qunit/no-conditional-assertions */
```
Not sure if there's a good way to avoid this 🤔 

The data structure and complexity is a bit high for `assert.step` and `assert.verifySteps`.

I'm pretty confident that `replayEvents` is simple enough that assertions are running when we expect them to.

gilest commented on test-app/tests/unit/system/upload-progress-test.ts:
```diff
@@ -0,0 +1,202 @@
+    replayEvents(assert, uploadFile, steps);
+  });
+
+  test('Firefox (from issue #1002)', function (this: TestContext, assert) {
```
https://user-images.githubusercontent.com/19845290/281305620-5bc45c54-8607-4c3a-88c3-c317c7cc1aee.png

gilest commented on test-app/tests/unit/system/upload-progress-test.ts:
```diff
@@ -0,0 +1,202 @@
+    replayEvents(assert, uploadFile, steps);
+  });
+
+  test('Chrome (from issue #1002)', function (this: TestContext, assert) {
```
https://user-images.githubusercontent.com/19845290/281305327-9240f787-1177-4ca4-b11f-e61f0f732c6a.png

gilest commented on test-app/tests/unit/system/upload-progress-test.ts:
```diff
@@ -0,0 +1,202 @@
+module('Unit | upload progress', function (hooks) {
+  setupTest(hooks);
+
+  test('Original (from issue #1002)', function (this: TestContext, assert) {
```
https://user-images.githubusercontent.com/36919/280431274-c69149b9-9918-4feb-9d5d-699a6dc59a3c.png

gilest commented:
@mkszepp please review ☝🏻

mkszepp commented:
@gilest looks good, added two comments in code...
We should also add a test in which `loadend` brings `total: 0`, because this would be happen, when sombody returns only status 2xx with no content

gilest commented:
> We should also add a test in which `loadend` brings `total: 0`, because when sombody returns only status 2xx with no content it occurs

Ah yes – great idea. I have added a test for this.

> @gilest looks good, added two comments in code...

Thanks – not sure if your code comments got saved, I cannot see any...

I'm going to merge and release but feel free to open a tidy-up PR with your suggestions
